File: opensource/data/src/main/java/com/sos/scheduler/engine/data/folder/TypedPath.java
package com.sos.scheduler.engine.data.folder;

import java.io.File;

public class TypedPath extends AbsolutePath {
    private final FileBasedType typ;

    protected TypedPath(FileBasedType typ, AbsolutePath path) {
        super(path);
        this.typ = typ;
    }

    public final File file(File baseDirectory) {
        return typ.file(baseDirectory, this);
    }

    public final FileBasedType typ() {
        return typ;
    }

//    public int hashCode() {
//        return 31*typ.hashCode() + super.hashCode();
//    }

    @Override public boolean equals(Object o) {
        return o instanceof TypedPath && eq((TypedPath)o);
    }

    private boolean eq(TypedPath o) {
        return typ == o.typ && super.equals(o);
    }

    @Override public String toString() {
        return typ+" "+super.toString();
    }

    public static TypedPath of(FileBasedType t, String absolutePath) {
        return of(t, new AbsolutePath(absolutePath));
    }

    public static TypedPath of(FileBasedType t, AbsolutePath absolutePath) {
        switch (t) {
            case job: return new JobPath(absolutePath);
            case jobChain: return new JobChainPath(absolutePath);
            case folder:
            case lock:
            case order:
            case processClass:
            case schedule:
                return new TypedPath(t, absolutePath);  // Bis die anderen ...Path implementiert sind.
        }
        throw new IllegalArgumentException();  // Javac will das
    }
}

File: opensource/data/src/main/scala/com/sos/scheduler/engine/data/folder/TypedPath.scala
package com.sos.scheduler.engine.data.folder

import java.io.File
import scala.Predef.String

trait TypedPath
extends AbsolutePath {

  protected def typ: FileBasedType

  def file(baseDirectory: File): File =
    new File(baseDirectory, relativeFilePath)

  /** @return Relativer Pfad mit Schrägstrich beginnend. Großschreibung kann bei manchen Typen abweichen. */
  def relativeFilePath: String =
    if (typ eq FileBasedType.folder) string + "/"
    else string + "." + typ.cppName + ".xml"

  override def toString =
    s"$typ ${super.toString}"
}
